hal/api/toolchain.h

```diff
@@ -16,6 +16,214 @@
 #ifndef MBED_TOOLCHAIN_H
 #define MBED_TOOLCHAIN_H
 
+
+// Warning for unsupported compilers
+#if !defined(__GNUC__)   /* GCC        */ \
+ && !defined(__CC_ARM)   /* ARMCC      */ \
+ && !defined(__clang__)  /* LLVM/Clang */ \
+ && !defined(__ICCARM__) /* IAR        */
+#warning "This compiler is not yet supported."
+#endif
+
+
+// Attributes
+
+/** MBED_PACKED
+ *  Pack a structure, preventing any padding from being added between fields.
+ *
+ *  @code
+ *  #include "toolchain.h"
+ *
+ *  typedef struct {
+ *      char x;
+ *      int y;
+ *  } MBED_PACKED foo;
+ *  @endcode
+ */
+#ifndef MBED_PACKED
+#if defined(__ICCARM__)
+#define MBED_PACKED __packed
```

Comment:
I originally went this style since that is what existed before in toolchain.h with the PACKED keyword:
```
typedef struct { blabalbla } MBED_PACKED foo;
```

The issue is, as usual, IAR, which only supports attributes before declarations. However, these attributes can be before typedefed declarations, which allows the above to work for all three compilers (`__attribute__((packed))` effects the struct, `__packed` effects the typdef).

As an alternative, what is your thoughts on the following? This is supported by both the gcc and iar style of struct attributes:
```
MBED_PACKED(struct) hi { blablabla };
```